core/src/main/scala/kafka/admin/ReassignPartitionsCommand.scala

```diff
@@ -16,233 +16,1531 @@
     val adminZkClient = new AdminZkClient(zkClient)
-    val brokerMetadatas = adminZkClient.getBrokerMetadatas(rackAwareMode, Some(brokerListToReassign))
+    for (brokerId <- zkClient.getAllBrokersInCluster.map(_.id)) {
+      val configs = adminZkClient.fetchEntityConfig(ConfigType.Broker, brokerId.toString)
+      if (!brokerLevelThrottles.flatMap(throttle => Option(configs.remove(throttle))).isEmpty) {
+        adminZkClient.changeBrokerConfig(Seq(brokerId), configs)
+      }
+    }
+  }
 
-    val partitionsToBeReassigned = mutable.Map[TopicPartition, Seq[Int]]()
+  /**
+   * Clear the reassignment throttles for the specified topics.
+   *
+   * @param adminClient           The AdminClient to use.
+   * @param topics                The topics to clear the throttles for.
+   */
+  def clearTopicLevelThrottles(adminClient: Admin, topics: Set[String]): Unit = {
+    val configOps = new util.HashMap[ConfigResource, util.Collection[AlterConfigOp]]()
+    topics.foreach {
+      topicName => configOps.put(
+        new ConfigResource(ConfigResource.Type.TOPIC, topicName),
+        topicLevelThrottles.map(throttle => new AlterConfigOp(new ConfigEntry(throttle, null),
+          OpType.DELETE)).asJava)
+    }
+    adminClient.incrementalAlterConfigs(configOps).all().get()
+  }
+
+  /**
+   * Clear the reassignment throttles for the specified topics.
+   *
+   * @param zkClient              The ZooKeeper client to use.
+   * @param topics                The topics to clear the throttles for.
+   */
+  def clearTopicLevelThrottles(zkClient: KafkaZkClient, topics: Set[String]): Unit = {
+    val adminZkClient = new AdminZkClient(zkClient)
+    for (topic <- topics) {
+      val configs = adminZkClient.fetchEntityConfig(ConfigType.Topic, topic)
+      if (!topicLevelThrottles.flatMap(throttle => Option(configs.remove(throttle))).isEmpty) {
+        adminZkClient.changeTopicConfig(topic, configs)
+      }
+    }
+  }
+
+  /**
+   * The entry point for the --generate command.
+   *
+   * @param adminClient           The AdminClient to use.
+   * @param reassignmentJson      The JSON string to use for the topics to reassign.
+   * @param brokerListString      The comma-separated string of broker IDs to use.
+   * @param enableRackAwareness   True if rack-awareness should be enabled.
+   *
+   * @return                      A tuple containing the proposed assignment and the
+   *                              current assignment.
+   */
+  def generateAssignment(adminClient: Admin,
+                         reassignmentJson: String,
+                         brokerListString: String,
+                         enableRackAwareness: Boolean)
+                         : (Map[TopicPartition, Seq[Int]], Map[TopicPartition, Seq[Int]]) = {
+    val (brokersToReassign, topicsToReassign) =
+      parseGenerateAssignmentArgs(reassignmentJson, brokerListString)
+    val currentAssignments = getReplicaAssignmentForTopics(adminClient, topicsToReassign)
```

Comment:
Since these two `generateAssignment` methods are nearly identical, do you think it's worth trying to factor out the common parts? Maybe we could have a single `generateAssignment` that accepts an `Either[Admin, ZkClient]`?

Reply:
Good question.  The zookeeper code is going away soon, so I don't think it's worth trying to combine the functions.

I could write a function to factor out the "print" statements, but I'm not sure it's worth it since the main work is done by `calculateAssignment` and `formatAsReassignmentJson`, which are already shared code.

Reply:
Yea, probably not worth the effort in this case.